tests/test_analog.py

```diff
@@ -142,12 +143,20 @@ def test_hamiltonianevolution_with_types(
 ) -> None:
     n_qubits = 4
     hamevo = pyq.HamiltonianEvolution(H, t_evo, tuple([i for i in range(n_qubits)]))
+    assert len(hamevo._cache_hamiltonian_evo) == 0
     psi = pyq.uniform_state(n_qubits)
     psi_star = hamevo(psi)
     result = overlap(psi_star, psi)
     assert result.size() == (batch_size,)
     assert torch.allclose(result, target, rtol=RTOL, atol=ATOL)
 
+    # test cached
+    assert len(hamevo._cache_hamiltonian_evo) == 1
+    psi_star = hamevo(psi)
+    assert len(hamevo._cache_hamiltonian_evo) == 1
+    # check value has not changed
+    assert torch.allclose(result, target, rtol=RTOL, atol=ATOL)
```

Comment:
You need to recompute `result = overlap(psi_star, psi)` for this to make sense right?